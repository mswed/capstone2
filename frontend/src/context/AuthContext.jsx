import React, { useState, useEffect } from 'react';
import { jwtDecode } from 'jwt-decode';
import GrumpyApi from '../services/api';

// We create the context
const AuthContext = React.createContext();

/**
 * This is a logic only component that wraps our entire app and sets the
 * authentication context and its functions
 *
 * @param {Array}  children - Components to wrap
 * @returns {AuthContext.Provider} - with auth functions for easy access
 */

const AuthProvider = ({ children }) => {
  // We grab stuff from local storage if available
  const storedToken = localStorage.getItem('grumpy-token');
  const storedUser = localStorage.getItem('grumpy-user');

  // We set the token right away so we don't run in to authorization issues
  GrumpyApi.token = storedToken || null;

  // Initialize state
  const [token, setToken] = useState(storedToken);
  const [currentUser, setCurrentuser] = useState(storedUser);

  // Whenever the token changes (login/logout) we need to update
  // both the api module and localStorage
  useEffect(() => {
    if (token && token.trim !== '') {
      // We only set the token if it's not empty

      GrumpyApi.token = token;
      localStorage.setItem('grumpy-token', token || '');
      localStorage.setItem('grumpy-user', currentUser || '');
    } else {
      GrumpyApi.token = null;
      localStorage.removeItem('grumpy-token');
      localStorage.removeItem('grumpy-user');
    }
  }, [token]);

  // TODO:
  // // We also want to get the user favorites here
  // useEffect(() => {
  //   if (!token) return;
  //   async function getUserData() {
  //     const userData = await GrumpuyApi.getUser(currentUser);
  //     setFavorites(new Set(userData.favorites));
  //   }
  //   getUserData();
  // }, [token]);

  const login = async (username, password) => {
    try {
      const authToken = await GrumpyApi.login(username, password);
      setToken(authToken);
      const decoded = jwtDecode(authToken);
      setCurrentuser(decoded.username);
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  const register = async (username, password, firstName, lastName, email) => {
    try {
      const token = await GrumpyApi.register(username, password, firstName, lastName, email);
      logout();
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  const logout = () => {
    // TODO: Do I need to logout of the backend here too?
    setToken('');
  };

  // TODO: We want to add favorites here
  // const addFavorite = async (jobId) => {
  //   const res = await GrumpuyApi.favoriteFormat(currentUser, jobId);
  //   if (res) {
  //     setFavorites((currentFavorites) => currentFavorites.add(res.favorites));
  //   }
  // };
  //

  // TODO: get items that have already been favorited
  // const alreadyFavorited = (formatId) => {
  //   return favFormats.has(formatId);
  // };

  // Create the value object
  const value = {
    token,
    currentUser,
    login,
    register,
    logout,
  };

  // Wrap everything in the context
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export { AuthContext, AuthProvider };
