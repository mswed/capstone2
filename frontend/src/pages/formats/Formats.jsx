import { useState, useEffect } from 'react';
import { Container, Card, Row, Col, Button } from 'react-bootstrap';
import GrumpyApi from '../../services/api';
import Loading from '../../components/ui/Loading';
import AdvanceSearchForm from '../../components/forms/AdvanceSearchForm';
import FormatList from '../../features/formats/components/FormatList';

const Formats = () => {
  const [formats, setFormats] = useState([]);
  const [formatCount, setFormatCount] = useState(0);
  const [searchParams, setSearchParams] = useState({});
  // We need to check if we're still loading so we won't run into trying to
  // render a null state
  const [isLoading, setIsLoading] = useState(true);

  /**
   * Get all the formats in the database. Eventually this should have a limit
   * or pagination but for now we get everything. This is called by useEffect
   * if the searchPArams changes and is empty
   *
   * @returns {Array} all cameras in the database
   */

  const getAllFormats = async () => {
    try {
      const response = await GrumpyApi.getFormats();
      setFormats(response);
    } catch (error) {
      console.error('Error fetching cameras', error);
    } finally {
      setIsLoading(false);
    }
  };

  /**
   * Get found formats in the databse. This is called by useEffect
   * if the searchParams changes and is NOT empty
   *
   * @returns {Array} found cameras in the database
   */

  const findFormats = async () => {
    try {
      const response = await GrumpyApi.findFormats(searchParams);
      setFormats(response);
    } catch (error) {
      console.error('Error fetching cameras', error);
    } finally {
      setIsLoading(false);
    }
  };

  const handleClearSearch = () => {
    setSearchParams([]);
  };
  useEffect(() => {
    if (Object.keys(searchParams).length !== 0) {
      findFormats();
    } else {
      getAllFormats();
    }
    window.scrollTo(0, 0);
  }, [searchParams]);

  useEffect(() => {
    async function getFormatCount() {
      try {
        const res = await GrumpyApi.getStats();
        setFormatCount(res.formats);
      } catch (error) {
        console.error('Error fetching stats', error);
      }
    }
    getFormatCount();
  }, []);

  if (isLoading) {
    return <Loading />;
  }
  return (
    <Container>
      <Row>
        <Col md={9}>
          <FormatList formats={formats} showModel={true} />
          <Row>
            <div className="text-muted text-center">Showing {formats.length} formats</div>
          </Row>
          {formatCount !== formats.length && (
            <Row className="mt-4">
              <Button variant="outline-danger" onClick={handleClearSearch}>
                Clear
              </Button>
            </Row>
          )}
        </Col>
        <Col md={3}>
          <Card className="m-3 px-3 shadow-lg">
            <AdvanceSearchForm search={setSearchParams} />
          </Card>
        </Col>
      </Row>
    </Container>
  );
};

export default Formats;
