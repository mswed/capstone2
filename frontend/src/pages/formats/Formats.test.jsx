import { vi, it, expect, beforeEach } from 'vitest';
import { fireEvent, render, screen, waitFor, within } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Formats from './Formats.jsx';
import { AuthContext } from '../../context/AuthContext.jsx';

// Formats search scrolls up when results change so we need to mock
// scrollTo
global.scrollTo = vi.fn();

// Mock API calls
vi.mock('../../services/api', () => ({
  default: {
    getFormats: vi.fn(() =>
      Promise.resolve([
        {
          id: 1,
          camera: 1,
          cameraModel: 'Alexa 35',
          makeName: 'ARRI',
          imageFormat: '4.6K',
          imageAspect: '3:2',
          formatName: 'Open Gate',
          sensorWidth: 28.0,
          sensorHeight: 19.2,
          imageWidth: 4608,
          imageHeight: 3164,
          pixelAspect: 1.0,
          isAnamorphic: false,
          isDesqueezed: false,
          anamorphicSqueeze: null,
          filmbackWidth3de: 28.0,
          filmbackHeight3de: 19.2,
          isDownsampled: false,
          isUpscaed: false,
          codec: 'ARRIRAW',
          rawReordingAvailable: true,
          source: 1,
          notes: 'sample notes',
          makeNotes:
            '4.6K 3:2 Open Gate provides maximum image quality, resolution, and flexibility in post for many spherical and anamorphic lenses in an image area slightly larger than traditional Super 35 film specifications.',
          trackingWorkflow: 'Do this then that',
        },
        {
          id: 2,
          camera: 1,
          cameraModel: 'Alexa 35',
          makeName: 'ARRI',
          imageFormat: '4K',
          imageAspect: '16:9',
          sensorWidth: 28.0,
          sensorHeight: 15.7,
          imageWidth: 4096,
          imageHeight: 2304,
          pixelAspect: 1.0,
          isAnamorphic: false,
          isDesqueezed: false,
          anamorphicSqueeze: null,
          filmbackWidth3de: 28.0,
          filmbackHeight3de: 19.2,
          isDownsampled: true,
          isUpscaed: false,
          codec: 'ProRes',
          rawReordingAvailable: true,
          source: 1,
          notes: 'Downsampled from 4.6K 16:9',
          makeNotes:
            '4.6K 3:2 Open Gate provides maximum image quality, resolution, and flexibility in post for many spherical and anamorphic lenses in an image area slightly larger than traditional Super 35 film specifications.',
          trackingWorkflow: 'Do that then this',
        },
      ])
    ),
    findFormats: vi.fn(() =>
      Promise.resolve([
        {
          id: 3,
          camera: 3,
          cameraModel: 'KOMODO',
          makeName: 'RED',
          imageFormat: '6K',
          imageAspect: '2.4:1',
          sensorWidth: 27.03,
          sensorHeight: 11.4,
          imageWidth: 6144,
          imageHeight: 2592,
          codec: 'R3D',
          source: 5,
          notes: 'There is a discrapency in the documentation the height is either 2592 or 2574',
          pixelAspect: 1.0,
          isAnamorphic: false,
          isDesqueezed: false,
          anamorphicSqueeze: null,
          filmbackWidth3de: 28.0,
          filmbackHeight3de: 19.2,
          isDownsampled: true,
          isUpscaed: false,
          rawReordingAvailable: true,
          makeNotes: '10 out of 10, no notes',
          trackingWorkflow: 'Do that then this',
        },
      ])
    ),
    getStats: vi.fn(() =>
      Promise.resolve({
        projects: 30,
        cameras: 10,
        makes: 3,
        formats: 200,
      })
    ),
  },
}));

describe('Formats component', () => {
  /**
   * Render the componenet with authentication context
   *
   * @returns {Component} - Makes component wrapped in MemoryRouter and AuthContext
   */

  const renderWithAuth = (authValue = { token: 'fake-token', currentUser: 'testUser' }) => {
    return render(
      <MemoryRouter>
        <AuthContext.Provider value={authValue}>
          <Formats />
        </AuthContext.Provider>
      </MemoryRouter>
    );
  };

  beforeEach(() => {
    // Clear mocks
    vi.clearAllMocks();
  });

  it('renders without crashing', async () => {
    renderWithAuth();
    await screen.findAllByText('ARRI');
  });

  it('matches snapshot', async () => {
    const { asFragment } = renderWithAuth();

    await screen.findAllByText('ARRI');
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays the correct info', async () => {
    renderWithAuth();
    await screen.findAllByText('ARRI');

    const table = screen.getByRole('table');
    const rows = within(table).getAllByRole('row');

    await waitFor(() => {
      // Table headers
      expect(screen.getByRole('columnheader', { name: 'Make' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Model' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Format' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Resolution' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Filmback' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'PAR' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Anamorphic?' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Desqueezed?' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Action' })).toBeInTheDocument();

      // Table data
      const firstDataRow = rows[1]; // Skip header row
      expect(within(firstDataRow).getByText('ARRI')).toBeInTheDocument();
      expect(within(firstDataRow).getByText('Alexa 35')).toBeInTheDocument();
      expect(within(firstDataRow).getByText('4.6K 3:2 Open Gate')).toBeInTheDocument();
      expect(within(firstDataRow).getByText('4608 x 3164')).toBeInTheDocument();
      expect(within(firstDataRow).getByText('28mm x 19.2mm')).toBeInTheDocument();
    });
  });

  it('can search for formats', async () => {
    renderWithAuth();

    await screen.findAllByText('ARRI');

    // Fill out the required fields
    const searchButton = screen.getByRole('button', { name: /search/i });

    fireEvent.click(searchButton);

    await waitFor(() => {
      // Table headers
      expect(screen.getByRole('columnheader', { name: 'Make' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Model' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Format' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Resolution' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Filmback' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'PAR' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Anamorphic?' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Desqueezed?' })).toBeInTheDocument();
      expect(screen.getByRole('columnheader', { name: 'Action' })).toBeInTheDocument();

      // Table data
      expect(screen.getByText('RED')).toBeInTheDocument();
    });
  });
});
