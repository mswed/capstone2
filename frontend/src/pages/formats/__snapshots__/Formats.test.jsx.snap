// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Formats component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-9"
      >
        <div
          class="table-responsive"
        >
          <table
            class="small table table-sm table-striped table-hover"
          >
            <thead>
              <tr>
                <th>
                  Make
                </th>
                <th>
                  Model
                </th>
                <th>
                  Format
                </th>
                <th>
                  Resolution
                </th>
                <th>
                  Filmback
                </th>
                <th>
                  PAR
                </th>
                <th>
                  Anamorphic?
                </th>
                <th>
                  Desqueezed?
                </th>
                <th>
                  Action
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  ARRI
                </td>
                <td>
                  Alexa 35
                </td>
                <td>
                  4.6K 3:2 Open Gate
                </td>
                <td>
                  4608 x 3164
                </td>
                <td>
                  28mm x 19.2mm
                </td>
                <td>
                  1
                </td>
                <td
                  class="text-center"
                >
                  <span
                    class="text-danger"
                  >
                    ✗
                  </span>
                </td>
                <td
                  class="text-center"
                >
                  <span
                    class="text-danger"
                  >
                    ✗
                  </span>
                </td>
                <td>
                  <a
                    class="btn btn-outline-primary btn-sm"
                    data-discover="true"
                    href="/formats/1"
                  >
                    View
                  </a>
                </td>
              </tr>
              <tr>
                <td>
                  ARRI
                </td>
                <td>
                  Alexa 35
                </td>
                <td>
                  4K 16:9 
                </td>
                <td>
                  4096 x 2304
                </td>
                <td>
                  28mm x 15.7mm
                </td>
                <td>
                  1
                </td>
                <td
                  class="text-center"
                >
                  <span
                    class="text-danger"
                  >
                    ✗
                  </span>
                </td>
                <td
                  class="text-center"
                >
                  <span
                    class="text-danger"
                  >
                    ✗
                  </span>
                </td>
                <td>
                  <a
                    class="btn btn-outline-primary btn-sm"
                    data-discover="true"
                    href="/formats/2"
                  >
                    View
                  </a>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div
          class="row"
        >
          <div
            class="text-muted text-center"
          >
            Showing 2 formats
          </div>
        </div>
        <div
          class="mt-4 row"
        >
          <button
            class="btn btn-outline-danger"
            type="button"
          >
            Clear
          </button>
        </div>
      </div>
      <div
        class="col-md-3"
      >
        <div
          class="m-3 px-3 shadow-lg card"
        >
          <form
            class="text-start"
          >
            <div
              class="d-flex justify-content-between align-items-center"
            >
              <h5
                class="mb-0"
              >
                Filter Formats
              </h5>
              <button
                class="border-0 btn btn-outline-danger btn-sm"
                type="button"
              >
                X
              </button>
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="advanceSearchForm.makeInput"
              >
                Make
              </label>
              <input
                class="form-control"
                id="advanceSearchForm.makeInput"
                name="make"
                type="text"
                value="Any"
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="advanceSearchForm.modelInput"
              >
                Model
              </label>
              <input
                class="form-control"
                id="advanceSearchForm.modelInput"
                name="camera"
                type="text"
                value="Any"
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="advanceSearchForm.formatInput"
              >
                Format
              </label>
              <input
                class="form-control"
                id="advanceSearchForm.formatInput"
                name="format"
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="advanceSearchForm.sensorWidthInput"
              >
                Sensor Width
              </label>
              <input
                class="form-control"
                id="advanceSearchForm.sensorWidthInput"
                name="sensor_width"
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="advanceSearchForm.sensorHeightInput"
              >
                Sensor Height
              </label>
              <input
                class="form-control"
                id="advanceSearchForm.sensorHeightInput"
                name="sensor_height"
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="advanceSearchForm.imageWidthInput"
              >
                Image Width
              </label>
              <input
                class="form-control"
                id="advanceSearchForm.imageWidthInput"
                name="image_width"
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="advanceSearchForm.imageHeightInput"
              >
                Image Height
              </label>
              <input
                class="form-control"
                id="advanceSearchForm.imageHeightInput"
                name="image_height"
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="advanceSearchForm.pixelAspectInput"
              >
                Pixel Aspect
              </label>
              <input
                class="form-control"
                id="advanceSearchForm.pixelAspectInput"
                name="pixel_aspect"
                type="text"
                value="1"
              />
            </div>
            <div
              class="form-check form-switch"
            >
              <input
                class="form-check-input"
                id="isAnamorphic"
                name="is_anamorphic"
                type="checkbox"
              />
              <label
                class="form-check-label"
                for="isAnamorphic"
                title=""
              >
                Anamorphic?
              </label>
            </div>
            <div
              class="form-check form-switch"
            >
              <input
                class="form-check-input"
                id="isDesqueezed"
                name="is_desqueezed"
                type="checkbox"
              />
              <label
                class="form-check-label"
                for="isDesqueezed"
                title=""
              >
                Desqueezed?
              </label>
            </div>
            <div
              class="form-check form-switch"
            >
              <input
                class="form-check-input"
                id="isDownsampled"
                name="is_downsampled"
                type="checkbox"
              />
              <label
                class="form-check-label"
                for="isDownsampled"
                title=""
              >
                Downsampled?
              </label>
            </div>
            <div
              class="form-check form-switch"
            >
              <input
                class="form-check-input"
                id="isUpscaled"
                name="is_upscaled"
                type="checkbox"
              />
              <label
                class="form-check-label"
                for="isUpscaled"
                title=""
              >
                Upscaled?
              </label>
            </div>
            <button
              class="btn btn-primary w-100 my-3 btn btn-primary"
              type="submit"
            >
              Search
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
